[Marcus Choo] iP

marcusc55 opened the pull request:
# Duke Chatbot 

> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke is a chatbot written in **Java**! Duke helps you to remember and manage your tasks.

Duke
1. is easy to use
2. supports multiple types of tasks
   1. To-dos
   2. Deadlines
   3. Events

and Duke can help you to:
* Add tasks
* View tasks
* Remove tasks
* Mark tasks as done
* Show your tasks
* Search for tasks

You can download Duke from [here](https://github.com/marcusc55/ip/releases/)!

Features:
- [x] Managing tasks
- [x] Managing deadlines and events with dates
- [ ] GUI support (Coming soon!)

***
If you are a Java programmer, you can use it to practice Java too! Here's the `Launcher` method:
```java
public class Launcher { 
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

marcusc55 committed:
Add greet, echo, exit functionality

marcusc55 committed:
Add tasks and list functionality

- Duke now stores user tasks and displays tasks when given "list" command.
- Duke is now instanced.

marcusc55 committed:
Added Task class

Task class represents a user's tasks.

marcusc55 committed:
Added ability to mark tasks as done

Duke now marks a task as done when user inputs "done X", where X is the task number to be marked as done.

marcusc55 committed:
Add todo, deadline and event classes

- Todo class represents a task without any date/time
- Deadline class represents a task that needs to be done before a specific date/time
- Event class represents a task that starts and ends at a specific time

marcusc55 committed:
Add support for tracking todos, events, deadlines

- Add todos using the command "todo"
- Add deadlines using the command "deadline". Deadlines must include a due date after the "/by" keyword.
- Add events using the command "event". Events must include a time after the "/at" keyword.

marcusc55 committed:
Add automated text-UI testing

marcusc55 committed:
Add error handling

Duke now handles these errors: 
- Invalid command
- Done command without providing task number/invalid task number
- Any task command without providing a task description
- Deadline or event command without providing due date or time respectively

Test cases for automated text-UI testing updated

marcusc55 committed:
Add delete functionality

- Duke now deletes tasks with the keyword "delete" followed by the number of the task to be deleted.
- Automated text-UI tests updated.

marcusc55 committed:
Refactor Duke into packages

- Added packages for Duke and Duke.Task
- Refactor Main function into Main.java file

marcusc55 committed:
Abstract TaskList into separate class

Abstracted task list and related functions (e.g. add, delete, setDone) into separate TaskList class.

marcusc55 committed:
Add storage

- Duke now saves tasks to a text file whenever changes are made to the task list.
- Storage is abstracted into its own class.

marcusc55 committed:
Extract parser into Parser class

marcusc55 committed:
Extract commands into Command classes

All commands now inherit from a parent Command class.

marcusc55 committed:
Add date and times

Dates are now represented using LocalDateTime objects.

marcusc55 committed:
Merge branch 'branch-Level-7'

marcusc55 committed:
Update Storage class to use date objects

Storage class now works with LocalDateTime objects.

marcusc55 committed:
Merge branch 'branch-Level-8'

marcusc55 committed:
Extract user interaction into Ui class

marcusc55 committed:
Add JUnit tests

marcusc55 committed:
Add manifest

marcusc55 committed:
Add JavaDoc comments

marcusc55 committed:
Edit code to follow coding standard

marcusc55 committed:
Merge branch 'branch-A-CodingStandard'

marcusc55 committed:
Merge branch 'branch-A-JavaDoc'

marcusc55 committed:
Add find command

Users can now search their task lists using the 'find' command.

marcusc55 committed:
Merge branch 'branch-Level-9'

marcusc55 committed:
Merge branch 'branch-Level-8'

marcusc55 committed:
Fix command imports

marcusc55 committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

marcusc55 committed:
Refactor Main class into duke package

marcusc55 committed:
Merge branch 'branch-A-Gradle'

marcusc55 committed:
Add checkstyle configuration

marcusc55 committed:
Implement checkstyle suggestions

marcusc55 committed:
Merge branch 'branch-A-Checkstyle'

sw-boi commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+ */
+public class TaskList {
+    private static final String STORAGE_PATH = "data/duke.txt";
+    private List<Task> list = new ArrayList<>();
```
Maybe rename this List<Task> as "tasks"? To represent a list of Task objects.

marcusc55 replied:
Thanks for spotting this, will rename this.

sw-boi reviewed (commented):
LGTM. I really like how clean your code is. 😄 I aspire to code like this.

marcusc55 committed:
Add basic GUI support

Currently, GUI only supports echoing user's commands.

marcusc55 committed:
Add GUI and remove text UI

Duke commands now work properly in GUI

Tianqi-Zhu commented on src/main/java/duke/Main.java:
```diff
@@ -0,0 +1,12 @@
+
+public class Main {
+    /**
+     * Main class to start the Duke chatbot.
```
Should be an empty line between description and parameters.

marcusc55 replied:
Thanks for spotting this, fixed in a later commit!

Tianqi-Zhu commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,46 @@
+public abstract class Command {
+    public static final Command[] COMMANDS = new Command[] { new DeadlineCommand(), new DeleteCommand(),
+        new DoneCommand(), new EndCommand(), new EventCommand(), new FindCommand(), new ListCommand(),
+        new TodoCommand() };
```
Indentation for wrapped lines should be 8 spaces.

marcusc55 replied:
Thanks for spotting this, fixed in a later commit!

marcusc55 committed:
Merge branch 'branch-Level-10'

marcusc55 committed:
Replace Duke GUI photos

- Changed Duke GUI photos from template photos
- Removed text-ui-test files

marcusc55 committed:
Edit README

marcusc55 committed:
Improve code quality

marcusc55 committed:
Merge pull request #1 from marcusc55/branch-A-CodeQuality

Improve code quality

marcusc55 committed:
Use assertions

marcusc55 committed:
Merge pull request #2 from marcusc55/branch-A-Assertions

Use assertions

marcusc55 committed:
Add help command

Duke does not show what commands it accepts and what formats users should use to give commands.

It is difficult for user to know what commands Duke accepts.

By adding a "help" command that lists all of Duke's commands and parameters for its commands, users are able to easily see a list of Duke's commands and how to use them.

marcusc55 committed:
Merge branch 'branch-C-Help'

marcusc55 committed:
Set up CI with GitHub Actions

marcusc55 committed:
Fix bug in ParserTest

marcusc55 committed:
Set theme jekyll-theme-cayman

marcusc55 committed:
Add user guide and product photo

marcusc55 committed:
Fix exit command
